### ISSUE TITLE
Configurable Storage Class

### ISSUE THREAD

**Issue by jkoehler-redhat** [opendatahub-io/opendatahub-community#56]
This is a tracker for all the various bits we will need to track to complete the feature work to add ability to configure storage classes

# Requirements @jedemo please add all the requirements here.

Add requirements

# Individual Efforts

* UX: https://github.com/opendatahub-io/sig-ml-developer-experience/issues/44
    * Lead: Kyle Walker
* UI: https://github.com/opendatahub-io/odh-dashboard/issues/1270
    * Lead: @lucferbux 
* Dev: N/A
    * Lead: N/A
* QE: (add link)
    * Lead: TBD
* Doc: (add link)
    * Lead: TBD

**Comment by shalberd** [comment]
Also a topic in [odh data science pipelines operator ](https://github.com/opendatahub-io/data-science-pipelines-operator/issues/192)

**Comment by shalberd** [comment]
the key thing to keep in mind that it is not at all unusual to have Openshift Clusters without any default storage class set at cluster-level. 

a) So, having a way in Data Science Projects to set a default storageclass name via ODH dashboard config and odh dashboard config CR, evaluating that and setting that when creating the PVC, would be the first step.

b) A second step could be to list all storageclasses on workbench creation GUI.

@andrewballantyne @lucferbux  However, given that data science projects PVC creation do not work at all on clusters without [default storage class set in Openshift config](https://docs.openshift.com/container-platform/4.13/storage/container_storage_interface/persistent-storage-csi-sc-manage.html#overview), e.g. because the CSI driver does not allow it, would you agree that me making a PR for case a) first would be ok?

**Comment by andrewballantyne** [comment]
@shalberd We are aware of the limitations -- we were hoping to have avoided doing a hack to get around the default storage class issue for DS Projects -- but this feature got bumped a bit for other engagements. I'm a little weary of a "default storage" setting being done up today without properly following through -- as it will likely be throwaway code in its current form (likely we will merge this functionality in with this feature in the long run). 

Since I don't really have an idea when we will be able to get to this feature among our other immediate objectives... I'm not "against" a PR to do move this along for the time being. Not sure what is the most future proof idea to setting this up though, since this feature is still in the pre-design stage (architecturally). 

I'm taking it you have a more immediate need for a default storage class outside of the Jupyter Tile? I think we have a workaround due to JupyterHub's implementations.

**Comment by shalberd** [comment]
@andrewballantyne yes, I have an immediate or relatively immediate need to per-data-science-project, dynamically-sized PVCs with an explicit storageClassName. I have seen the UI and flow discussions. My suggestion would be to take the value from notebookController.storageClassName, if set.
     ```
   Must be able to configure storage classes other than the OpenShift default.
        Must be able to configure a 'default' storage for RHODS. This does not have to be the OpenShift default.
        Must be able to configure the name of the storage class as it should appear for users if exposed in the end user UI. Note: per Guillaume's comment below, the end user UI should show the admin configured display name as well as the real name. 
```

Meaning, flow-wise:

- add a variable or type defaultStorageClass with name, display name, if not already present, in odh dashboard code
- add logic to find out whethere any of the storage classes available at cluster-level is tagged as default ( a metadata annotation)
- if openshift default storageclass is set, assign that name and details to the variable. / type
- if no storageClass cluster-level default is set, assign the name from notebookController.storageClassName of CRD OdhDashboardConfig. That already exists, which is good.
- if notebookController.storageClassName is set, it override the default from cluster-level
- in the data science projects workbench page, add a dropdown list for storageClass, under PVC size, for example
- select by default in the dropdown the storageClass from variable or type defaultStorageClass, giving users the opportunity to override that default selection, though, i.e. making the drowndown selectable.
- add storageClassName in PVC assembly spec of workbenches, similar to what was already done at jupyter tile level before

Now, for my part, I could imagine doing anything that is not GUI-related, i.e. even the storageClass type array at typescript level. I just would not add a default-selected dropdown yet at workbench-level, so 
Would that be a feasible way for my PR?

**Comment by andrewballantyne** [comment]
Spoke with Sven and have created https://github.com/opendatahub-io/odh-dashboard/issues/1701 to implement a minor version of this feature while we do other tasks.
